mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala

```diff
@@ -213,32 +221,36 @@ class StringIndexer @Since("1.4.0") (
     val labelsArray = $(stringOrderType) match {
       case StringIndexer.frequencyDesc =>
         val sortFunc = StringIndexer.getSortFunc(ascending = false)
-        countByValue(dataset, inputCols).map { counts =>
+        val orgStrings = countByValue(dataset, inputCols).toSeq
+        ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
           counts.toSeq.sortWith(sortFunc).map(_._1).toArray
-        }
+        }.toArray
       case StringIndexer.frequencyAsc =>
         val sortFunc = StringIndexer.getSortFunc(ascending = true)
-        countByValue(dataset, inputCols).map { counts =>
+        val orgStrings = countByValue(dataset, inputCols).toSeq
+        ThreadUtils.parmap(orgStrings, "sortingStringLabels", 8) { counts =>
```

Comment:
So the issue here isn't so much the size of the sort (number of labels) but number of columns? so it's not like sorting with Spark helps. I think this is reasonable. This will use the default thread pool, but that's probably not going to be an issue.

Reply:
Yeah, the scale of input columns of StringIndexer is the issue.